### PR TITLE
Two round key exchange

### PR DESCRIPTION
**Pull request by lsolodkova**
Initial functionality. Added some stubs for ZKP, pin_key is something derived from PIN - our low-entropy secret.

### CODE REVIEW

**Review by OTFlorian — DISMISSED** [review]

File: applet/src/main/java/jpake/jpakePassivePayload.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        output[currStart] = (byte) publicV3Encoded.length;
+        currStart += 1;
+        Util.arrayCopyNonAtomic(publicV3Encoded, (short) 0, output, currStart,(short)publicV3Encoded.length);
+        currStart += publicV3Encoded.length;
+        output[currStart] = (byte)result3.length;
+        currStart +=1;
+        Util.arrayCopyNonAtomic(result3, (short) 0, output, currStart,(short)result3.length);
+        return output;
+    }
+    public static jpakePassivePayload fromBytes(byte[] input){
```
Comment (mvondracek) [inline applet/src/main/java/jpake/jpakePassivePayload.java]: please add `throws` clause for cases when `input` is not a valid representation of `jpakePassivePayload `

File: applet/src/main/java/jpake/jpakeActiveFirstPayload.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        currStart += 1;
+        Util.arrayCopyNonAtomic(publicV2Encoded, (short) 0, output, currStart,(short)publicV2Encoded.length);
+        currStart += publicV2Encoded.length;
+        output[currStart] = (byte)result2.length;
+        currStart +=1;
+        Util.arrayCopyNonAtomic(result2, (short) 0, output, currStart,(short)result2.length);
+        return output;
+    }
+
+    public static jpakeActiveFirstPayload fromBytes(byte[] input){
```
Comment (mvondracek) [inline applet/src/main/java/jpake/jpakeActiveFirstPayload.java]: Please add `throws` clause for cases when `input` is not a valid representation of `jpakePassivePayload`.

File: applet/src/main/java/jpake/jpakeActiveSecondPayload.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        currStart += 1;
+        Util.arrayCopyNonAtomic(encPublicV, (short) 0, output, currStart, (short)encPublicV.length);
+        currStart += encPublicV.length;
+        output[currStart] = (byte)encResult.length;
+        currStart += 1;
+        Util.arrayCopyNonAtomic(encResult, (short) 0, output, currStart, (short)encResult.length);
+        return output;
+    }
+
+    public static jpakeActiveSecondPayload fromBytes(byte[] input){
```
Comment (mvondracek) [inline applet/src/main/java/jpake/jpakeActiveSecondPayload.java]: Please add `throws` clause for cases when `input` is not a valid representation of `jpakePassivePayload`.

File: applet/src/main/java/jpake/jpakePassivePayload.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        byte[] encPublicV3 = new byte[pubV3len];
+        Util.arrayCopyNonAtomic(input, currStart, encPublicV3, (short)0, pubV3len);
+        ECPoint publicV3 = jpakeActor.curve.decodePoint(encPublicV3);
+        currStart += pubV3len;
+        short result3len = input[currStart];
+        currStart += 1;
+        byte[] encResult3 = new byte[result3len];
+        Util.arrayCopyNonAtomic(input, currStart, encResult3, (short)0, result3len);
+        BigInteger result3 =  new BigInteger(encResult3);
+
```
Comment (mvondracek) [inline applet/src/main/java/jpake/jpakePassivePayload.java]: Check if `input.length` is `currStart + result3len` to see if provided `input` is **exactly** one `jpakePassivePayload` instance and nothing more.

Comment (mvondracek) [reply applet/src/main/java/jpake/jpakePassivePayload.java]: @lsolodkova

Comment (mvondracek) [reply applet/src/main/java/jpake/jpakePassivePayload.java]: Also relevant for other `fromBytes` methods.

File: applet/src/test/java/applet/jpakeActiveFirstPayloadTest.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        boolean G2same = afp.getG2().equals(restored.getG2());
+        boolean pubA1same = afp.getZKPx1().getPublicA().equals(restored.getZKPx1().getPublicA());
+        boolean pubV1same = afp.getZKPx1().getPublicV().equals(restored.getZKPx1().getPublicV());
+        boolean res1same = (afp.getZKPx1().getResult().compareTo(restored.getZKPx1().getResult()) == 0);
+        boolean pubA2same = afp.getZKPx2().getPublicA().equals(restored.getZKPx2().getPublicA());
+        boolean pubV2same = afp.getZKPx2().getPublicV().equals(restored.getZKPx2().getPublicV());
+        boolean res2same = (afp.getZKPx2().getResult().compareTo(restored.getZKPx2().getResult()) == 0);
+
+        if(!G1same || !G2same || !pubA1same || !pubV1same || !res1same || !pubA2same || !pubV2same || !res2same){
+            throw new EmIllegalStateException();
```
Comment (mvondracek) [inline applet/src/test/java/applet/jpakeActiveFirstPayloadTest.java:34]: you could use `Assert.assertEquals`, but this is fine too

**Review by mvondracek — CHANGES_REQUESTED** [review]
Please add checks as I mentioned in comments.

File: applet/src/main/java/jpake/jpakePassivePayload.java (hunk 1 above)
Comment (lsolodkova) [reply applet/src/main/java/jpake/jpakePassivePayload.java]: Done.

File: applet/src/main/java/jpake/jpakeActiveFirstPayload.java (hunk 2 above)
Comment (lsolodkova) [reply applet/src/main/java/jpake/jpakeActiveFirstPayload.java]: Done.

File: applet/src/main/java/jpake/jpakeActiveSecondPayload.java (hunk 3 above)
Comment (lsolodkova) [reply applet/src/main/java/jpake/jpakeActiveSecondPayload.java]: Done.

File: applet/src/main/java/jpake/jpakePassivePayload.java (hunk 4 above)
Comment (lsolodkova) [reply applet/src/main/java/jpake/jpakePassivePayload.java]: Done.

**Review by OTFlorian — DISMISSED** [review]

**Review by mvondracek — DISMISSED** [review]

**Review by mvondracek — APPROVED** [review]

**Pull request merged**